File: library/src/main/java/com/shxhzhxx/sdk/activity/ForResultActivity.java
package com.shxhzhxx.sdk.activity;

import android.content.Intent;
import android.os.Bundle;
import androidx.core.app.ActivityCompat;
import android.util.SparseArray;

public abstract class ForResultActivity extends PermissionRequestActivity {
    public abstract static class ResultListener {
        public void onResult(int resultCode, Intent data) {
        }
    }

    public SparseArray<ResultListener> mActivityResultCallback = new SparseArray<>();

    public void startActivityForResult(Intent intent, ResultListener listener) {
        startActivityForResult(intent, listener, null);
    }

    public void startActivityForResult(Intent intent, ResultListener listener, Bundle options) {
        if (listener == null)
            return;
        int requestCode = getRequestCode(mActivityResultCallback);
        mActivityResultCallback.put(requestCode, listener);
        ActivityCompat.startActivityForResult(this, intent, requestCode, options);
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        if (mActivityResultCallback.indexOfKey(requestCode) >= 0) {
            ResultListener listener = mActivityResultCallback.get(requestCode);
            mActivityResultCallback.delete(requestCode);
            listener.onResult(resultCode, data);
        }
    }
}

File: library/src/main/java/com/shxhzhxx/sdk/activity/ForResultActivity.kt
package com.shxhzhxx.sdk.activity

import android.app.Activity
import android.content.Intent
import android.os.Bundle
import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProviders
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.suspendCancellableCoroutine
import kotlin.coroutines.resume
import kotlin.coroutines.resumeWithException

abstract class ForResultActivity : PermissionRequestActivity() {
    private class MyViewModel : ViewModel() {
        val resultCallbacks = mutableMapOf<Int, (Int, Intent?) -> Unit>()
    }

    private val vm by lazy { ViewModelProviders.of(this).get(MyViewModel::class.java) }

    fun startActivityForResult(intent: Intent, onResult: (resultCode: Int, data: Intent?) -> Unit) {
        startActivityForResult(intent, vm.resultCallbacks.add(onResult))
    }

    suspend fun startActivityForResultCoroutine(intent: Intent, onFailure: (() -> Unit)? = null): Intent? =
            try {
                suspendCancellableCoroutine { continuation ->
                    startActivityForResult(intent) { resultCode, data ->
                        if (resultCode == Activity.RESULT_OK)
                            continuation.resume(data)
                        else
                            continuation.resumeWithException(CancellationException())
                    }
                }
            } catch (e: CancellationException) {
                onFailure?.invoke()
                throw e
            }

    final override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        vm.resultCallbacks.remove(requestCode)?.invoke(resultCode, data)
    }


    @Deprecated("Use callback version to avoid conflict.", ReplaceWith("startActivityForResult(intent){resultCode,data->}"))
    override fun startActivityForResult(intent: Intent, requestCode: Int) {
        super.startActivityForResult(intent, requestCode)
    }

    @Deprecated("Use callback version to avoid conflict.", ReplaceWith("startActivityForResult(intent){resultCode,data->}"))
    override fun startActivityForResult(intent: Intent, requestCode: Int, options: Bundle?) {
        super.startActivityForResult(intent, requestCode, options)
    }
}
